java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/SlidingWindowUtils.java

```diff
@@ -61,73 +71,86 @@ public class SlidingWindowUtils {
      * @param stepSizeMs The step size of the sliding window in milliseconds.
      * @param aggregationFieldsDescriptor The descriptor of the aggregation field in the sliding
      *     window.
+     * @param zeroValuedRow If the zeroValuedRow is not null, the sliding window will output a row
+     *     with default value when the window is empty. The zeroValuedRow contains zero values of
+     *     the all the fields, except row time field and key fields.
+     * @param skipSameWindowOutput Whether to output if the sliding window output the same result.
      */
-    public static Table applySlidingWindowKeyedProcessFunction(
+    public static Table applySlidingWindow(
             StreamTableEnvironment tEnv,
             Table table,
             String[] keyFieldNames,
             String rowTimeFieldName,
             long stepSizeMs,
-            AggregationFieldsDescriptor aggregationFieldsDescriptor) {
+            AggregationFieldsDescriptor aggregationFieldsDescriptor,
+            Row zeroValuedRow,
+            boolean skipSameWindowOutput) {
+        ResolvedSchema schema = table.getResolvedSchema();
+        DataStream<Row> stream =
+                tEnv.toChangelogStream(
+                        table,
+                        Schema.newBuilder().fromResolvedSchema(schema).build(),
+                        ChangelogMode.all());
+
+        final OutputTag<Row> outputTag = new OutputTag<Row>("late-data-side-output") {};
+
+        TimeZone timeZone = TimeZone.getTimeZone(tEnv.getConfig().getLocalTimeZone());
+
+        SingleOutputStreamOperator<Row> outputStream =
+                stream.keyBy(new RowKeySelector(keyFieldNames))
+                        .window(
+                                TumblingEventTimeWindows.of(
+                                        Time.milliseconds(stepSizeMs),
+                                        getModdedOffset(stepSizeMs, -timeZone.getRawOffset())))
```

Comment:
We are not handling daylight saving. Should we add a TODO as a reminder?

Reply:
According to offline discussion, Feathub may not need to handle daylight saving. We may assume each day have 24 hours.